File: Platforms/Windows/Reign.Video.D3D9/ShaderResource.cpp
#include "pch.h"
#include "ShaderResource.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Properties
	string^ ShaderResource::Name::get() {return name;}
	#pragma endregion

	#pragma region Contructors
	ShaderResource::ShaderResource(VideoI^ video, int vertexIndex, int pixelIndex, string^ name)
	{
		this->video = (Video^)video;

		this->vertexIndex = vertexIndex;
		this->pixelIndex = pixelIndex;

		this->name = name;
		Apply = gcnew ApplyFunc(this, &ShaderResource::setNothing);
	}
	#pragma endregion

	#pragma region Methods
	void ShaderResource::setNothing()
	{
		// Place holder.
	}

	void ShaderResource::setTexture2D()
	{
		if (vertexIndex != -1) video->Device->SetTexture(D3DVERTEXTEXTURESAMPLER0 + vertexIndex, resource);
		if (pixelIndex != -1) video->Device->SetTexture(pixelIndex, resource);
	}

	void ShaderResource::Set(Texture2DI^ resource)
	{
		Texture2D^ texture = (Texture2D^)resource;
		if (vertexIndex != -1) video->currentVertexTextures[vertexIndex] = texture;
		if (pixelIndex != -1) video->currentPixelTextures[pixelIndex] = texture;
		this->resource = texture->Texture;
		Apply = gcnew ApplyFunc(this, &ShaderResource::setTexture2D);
	}

	void ShaderResource::Set(Texture3DI^ resource)
	{
		/*Texture2D^ texture = (Texture2D^)resource;
		if (vertexIndex != -1) video->currentVertexTextures[vertexIndex] = texture;
		if (pixelIndex != -1) video->currentPixelTextures[pixelIndex] = texture;
		this->resource = texture->Texture;
		Apply = gcnew ApplyFunc(this, &ShaderResource::setTexture2D);*/
	}
	#pragma endregion
}
}
}

File: Platforms/Windows/Reign.Video.D3D9/Shaders/ShaderResource.cs
using System;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class ShaderResource : ShaderResourceI
	{
		#region Properties
		public delegate void ApplyMethod();
		public ApplyMethod Apply;
		private WeakReference resource;

		public string Name {get; private set;}
		private ShaderResourceCom com;
		private Video video;
		private int vertexIndex, pixelIndex;
		#endregion

		#region Constructors
		public ShaderResource(Video video, string name, int vertexIndex, int pixelIndex)
		{
			Name = name;
			this.video = video;
			this.vertexIndex = vertexIndex;
			this.pixelIndex = pixelIndex;

			com = new ShaderResourceCom(video.com, vertexIndex, pixelIndex);
			Apply = setNothing;
			resource = new WeakReference(null);
		}
		#endregion

		#region Methods
		private void setNothing()
		{
			// Place Holder...
		}

		private void setTexture2D()
		{
			com.SetTexture2D(((Texture2D)resource.Target).com);
		}

		public void Set(Texture2DI resource)
		{
			this.resource.Target = resource;
			if (vertexIndex != -1) video.currentVertexTextures[vertexIndex] = (Texture2D)resource;
			if (pixelIndex != -1) video.currentPixelTextures[pixelIndex] = (Texture2D)resource;
			Apply = setTexture2D;
		}

		public void Set(Texture3DI resource)
		{
			throw new NotImplementedException();
		}
		#endregion
	}
}
